core/src/main/scala/kafka/server/DelayedFetch.scala

```diff
@@ -111,7 +116,7 @@ class DelayedFetch(delayMs: Long,
     }
 
     // Case D
-    if (accumulatedSize >= fetchMetadata.fetchMinBytes)
+    if (accumulatedSize >= fetchMetadata.fetchMinBytes || (containsThrottledPartitions && !quota.isQuotaExceededBy(0)))
```

Comment:
Does this guarantee that the minBytes requirement is met? For example, suppose that all replicas are throttled and quota is not exceeded at this point. However, when we call forceComplete(0, we could get fewer bytes than minBytes and return the fetch response prematurely.

Same question in case C on line 94. If the fetch offset is on an old segment from the last segment, we force a return immediately. However, it could be that the quota is violated and we will return an empty response.